File: pyspark_learn_project/structured_api/rdds/BasicsOfRdd.py
from common_utils.sparkUtils import get_spark_session


class BasicOfRdd:
    """
    RDD: resilient distributed dataset low level api of spark
    when to use?
    1. tight control over data (/b/c of object of row)
    2. custom distributed variable
    3. to maintain some legacy code base
     RDD represents an immutable, partitioned collection of records that can be operated
on in parallel. Unlike DataFrames though, where each record is a structured row containing
fields with a known schema, in RDDs the records are just Java, Scala, or Python objects of the
programmer’s choosing
    """

    def demo(self):
        spark = get_spark_session()
        num_rdd = spark.range(500).rdd
        num_rdd1 = num_rdd.map(lambda x: x[0])
        print(num_rdd.take(5))

        # From local collection
        myCollection = "Spark The Definitive Guide : Big Data Processing Made Simple".split(" ")
        coll_rdd = spark.sparkContext.parallelize(myCollection, 2)
        print(coll_rdd.count())
        print(coll_rdd.collect())
        coll_rdd = coll_rdd.map(lambda x: x)
        print(coll_rdd.collect())

        print(coll_rdd.distinct().count())

    def filter_rdd(self):
        spark = get_spark_session()
        my_collection = "Spark The Definitive Guide : Big Data Processing Made Simple".split(" ")
        coll_rdd = spark.sparkContext.parallelize(my_collection, 2)

        def start_wit(individual):
            return individual.startswith("S")

        filtered_coll_rdd = coll_rdd.filter(lambda x: start_wit(x))
        print(filtered_coll_rdd.collect())

        map_coll_rdd = coll_rdd.map(lambda x: (x, x[0], start_wit(x)))
        print(map_coll_rdd.collect())

        # flattern the output of map with char by char
        flat_map_coll_rdd = coll_rdd.flatMap(lambda x: x)
        print(flat_map_coll_rdd.collect())

        sorted_coll_rdd = coll_rdd.sortBy(lambda x: len(x), ascending=True)
        print(sorted_coll_rdd.collect())

        # Action

        def reducer(left, right):
            if len(left) > len(right):
                return left
            else:
                return right

        word_reduced_rdd = coll_rdd.reduce(lambda x, y: reducer(x, y))
        print(word_reduced_rdd)

        word = spark.range(1, 34).rdd
        print(word.first())
        print(word.min())
        print(word.max())

        word.saveAsTextFile("C:\\Users\\sujee\\pydev\\pyspark_learn_project\\resources\\words_rdd")
        """To set a compression codec, we must import the proper codec from Hadoop. You can find these
in the org.apache.hadoop.io.compress library:"""

    def caching_and_storage_level(self):
        spark = get_spark_session()
        my_collection = "Spark The Definitive Guide : Big Data Processing Made Simple".split(" ")
        coll_rdd = spark.sparkContext.parallelize(my_collection, 2)

        print(f"checkpoint: {coll_rdd.getCheckpointFile()},"
              f" storage_level:{coll_rdd.getStorageLevel()},"
              f"num_of_partition:{coll_rdd.getNumPartitions()}")
        """org.apache.spark.storage.StorageLevel"""
        coll_rdd.cache()
        spark.sparkContext.setCheckpointDir("C:\\Users\\sujee\\pydev\\pyspark_learn_project\\resources\\checkping_dir")
        coll_rdd = coll_rdd.repartition(4)
        coll_rdd.checkpoint()

        print(f"checkpoint: {coll_rdd.getCheckpointFile()},"
              f" storage_level:{coll_rdd.getStorageLevel()},"
              f"num_of_partition:{coll_rdd.getNumPartitions()}")

    def key_value_based_rdd(self):
        spark = get_spark_session()
        my_coll = "Spark The Definitive Guide : Big Data Processing Made Simple" \
            .split(" ")
        words = spark.sparkContext.parallelize(my_coll, 2)

        pair_rdd_word = words.map(lambda x: (x.lower(), 1))
        words_key_by = words.keyBy(lambda x: x.lower()[0])

        print(words_key_by.collect())
        print(pair_rdd_word.collect())

        words_key_by = words_key_by.mapValues(lambda x: x.upper())

        print(words_key_by.collect())

        flat_map_value = words_key_by.flatMapValues(lambda x: x.upper())
        print(flat_map_value.collect())

        print(words_key_by.keys().collect())
        print(words_key_by.values().collect())

        print(words_key_by.lookup('S'))

    def rdd_operation_on_file(self):
        spark = get_spark_session()
        flight_path = "../../resources/spark_input/flight_data/2015-summary.csv"
        readme_path = "../../resources/spark_input/README.txt"
        flight_path_rdd = spark.sparkContext.textFile(flight_path)
        readme_path_rdd = spark.sparkContext.textFile(readme_path)
        header = flight_path_rdd.first()

        # print(flight_path_rdd.collect())
        flight_path_rdd_map = flight_path_rdd.filter(lambda x: x != header).map(lambda row: row.split(","))
        # print(flight_path_rdd_map.take(5))
        # print(flight_path_rdd_map.map(lambda x: x[0]).collect())

        counts_rdd = flight_path_rdd_map.filter(lambda x: x[0] == 'United States')
        # print(counts_rdd.collect())

        # print(readme_path_rdd.collect())

        splitted = readme_path_rdd.flatMap(lambda w: w)
        # print(splitted.take(10))
        print(flight_path_rdd_map.collect())
        print(flight_path_rdd_map.groupBy(lambda x: x[0]).map(lambda x: (x[0], x[1].map(lambda x: x[2]))).take(5))

    def test(self):
        spark = get_spark_session()
        path = "C:\\Users\\sujee\\pydev\\pyspark_learn_project\\resources\\mystartups.csv"

        startup_rdd = spark.sparkContext.textFile(path)
        header = startup_rdd.first()
        startup_rdd = startup_rdd.filter(lambda row: row != header).map(lambda x: x.split(","))
        print(startup_rdd.take(4))
        val = startup_rdd.map(lambda x: float(x[0])).sum()
        print(val)
        print(
            startup_rdd.map(lambda x: groupByKey(lambda x: x[3]).take(5)
                            )


def basic_of_rdd():
    obj = BasicOfRdd()
    # obj.key_value_based_rdd()
    # obj.rdd_operation_on_file()
    #obj.test()


if __name__ == '__main__':
    basic_of_rdd()

File: learnModule/databricks_code/007/LowLevelApi/BasicsOfRdd.scala
// Databricks notebook source
// MAGIC %md
// MAGIC There are times when higher-level manipulation will not meet the business or engineering problem you are trying to solve. For those cases, you
// MAGIC might need to use Spark’s lower-level APIs
// MAGIC
// MAGIC **There are two sets of low-level APIs**
// MAGIC 1. one for manipulating distributed data (RDDs)
// MAGIC 2. another for distributing and manipulating distributed shared variables (broadcast variables and
// MAGIC accumulators).
// MAGIC
// MAGIC
// MAGIC
// MAGIC **When to Use the Low-Level APIs?**
// MAGIC 1. need very tight control over physical data placement across the cluster/partioner/skewness
// MAGIC 2. maintain some legacy codebase 
// MAGIC 3. custome shared variable manupulation
// MAGIC
// MAGIC
// MAGIC **What is RDD?**
// MAGIC  RDD represents an immutable, partitioned collection of records that can be operated
// MAGIC on in parallel. Unlike DataFrames though, where each record is a structured row containing
// MAGIC fields with a known schema, in RDDs the records are just Java, Scala, or Python objects of the
// MAGIC programmer’s choosing.
// MAGIC

// COMMAND ----------

// MAGIC %md
// MAGIC **each RDD is characterized by five main properties:**
// MAGIC 1. A list of partitions
// MAGIC 2. A function for computing each split
// MAGIC 3. A list of dependencies on other RDDs
// MAGIC 4. Optionally, a Partitioner for key-value RDDs (e.g., to say that the RDD is hashpartitioned)
// MAGIC 5. Optionally, a list of preferred locations on which to compute each split (e.g., block
// MAGIC locations for a Hadoop Distributed File System [HDFS] fil

// COMMAND ----------

// MAGIC %run ./commonRdd

// COMMAND ----------

rangeOfNumbers.toDF().rdd.map(rowObject => rowObject.getLong(0)).take(5)

// COMMAND ----------

// MAGIC %md
// MAGIC ###Transformation on RDD
// MAGIC
// MAGIC 1. distinct: take the distinct value
// MAGIC 2. filter: filter the row
// MAGIC 3. map: *Specify a function that returns the value that you want, given the correct input. You then apply that, record by record*
// MAGIC 4. flatMap: *flatMap provides a simple extension of the map function we just looked at. Sometimes, each current row should return multiple rows*
// MAGIC 5. sortBy: words.sortBy(row => row.length * -1)
// MAGIC
// MAGIC ###Action
// MAGIC 1. collect: *collect the data on the driver*
// MAGIC 2. count:
// MAGIC 3. countByValue: *words.countByValue() You should use this method only if the resulting map is expected to be small because the entire thing is loaded into the driver’s memory.*
// MAGIC 3. reduce:  *words.reduce((x,y) => if(x.length > y.length) x else y)  //longest words*
// MAGIC 4. redcuceByKey: *require key.value pair
// MAGIC 5. *words.first()*
// MAGIC 6. min : *spark.sparkContext.parallelize(1 to 20).min()*
// MAGIC 7. max: *spark.sparkContext.parallelize(1 to 20).max()*
// MAGIC 8. words.saveAsTextFile("file:/tmp/bookTitle")
// MAGIC
// MAGIC

// COMMAND ----------

words.collect

// COMMAND ----------

words.distinct.count

// COMMAND ----------

//filter the RDD to keep only the words that begin with the letter “S”:
def LetterStartWith(abc:String):Boolean= abc.startsWith("S")

words.filter(rowObject => LetterStartWith(rowObject)).collect

// COMMAND ----------

val words2 = words.map(word => (word, word(0), word.startsWith("S")))

words2.filter(rowObject => rowObject._3 == true).take(5)

// COMMAND ----------

//flatMap
words.flatMap(row => row).collect

// COMMAND ----------

words.sortBy(row => row.length * -1).collect

// COMMAND ----------

val fiftyFiftySplit = words.randomSplit(Array[Double](0.5, 0.5))

fiftyFiftySplit(1).collect()

// COMMAND ----------

words.reduce((x,y) => if(x.length > y.length) x else y)  //longest words

// COMMAND ----------

words.countByValue() ///scala.collection.Map[String,Long]

// COMMAND ----------



// COMMAND ----------

// MAGIC %md
// MAGIC ###Caching
// MAGIC
// MAGIC Using cache() and persist() methods, Spark provides an optimization mechanism to store the intermediate computation of an RDD, DataFrame, and Dataset so they can be reused in subsequent actions(reusing the RDD, Dataframe, and Dataset computation results).
// MAGIC  
// MAGIC   cache() method default saves it to memory (MEMORY_ONLY) 
// MAGIC   whereas persist() method is used to store it to the user-defined storage level. 
// MAGIC
// MAGIC 1) *persist() : Dataset.this.type*
// MAGIC 2) *persist(newLevel : org.apache.spark.storage.StorageLevel) : Dataset.this.type*
// MAGIC
// MAGIC
// MAGIC `
// MAGIC MEMORY_ONLY – **This is the default behavior of the RDD cache()** method and stores the RDD or DataFrame as deserialized objects to JVM memory. When there is no enough memory available it will not save DataFrame of some partitions and these will be re-computed as and when required. This takes more memory. but unlike RDD, this would be slower than MEMORY_AND_DISK level as it recomputes the unsaved partitions and recomputing the in-memory columnar representation of the underlying table is expensive
// MAGIC
// MAGIC MEMORY_ONLY_SER – This is the same as MEMORY_ONLY but the difference being it stores RDD as serialized objects to JVM memory. It takes lesser memory (space-efficient) then MEMORY_ONLY as it saves objects as serialized and takes an additional few more CPU cycles in order to deserialize.
// MAGIC
// MAGIC MEMORY_ONLY_2 – Same as MEMORY_ONLY storage level but replicate each partition to two cluster nodes.
// MAGIC
// MAGIC MEMORY_ONLY_SER_2 – Same as MEMORY_ONLY_SER storage level but replicate each partition to two cluster nodes.
// MAGIC
// MAGIC MEMORY_AND_DISK – **This is the default behavior of the DataFrame or Dataset**. In this Storage Level, The DataFrame will be stored in JVM memory as a deserialized object. When required storage is greater than available memory, it stores some of the excess partitions into the disk and reads the data from the disk when required. It is slower as there is I/O involved.
// MAGIC
// MAGIC MEMORY_AND_DISK_SER – This is the same as MEMORY_AND_DISK storage level difference being it serializes the DataFrame objects in memory and on disk when space is not available.
// MAGIC
// MAGIC MEMORY_AND_DISK_2 – Same as MEMORY_AND_DISK storage level but replicate each partition to two cluster nodes.
// MAGIC
// MAGIC MEMORY_AND_DISK_SER_2 – Same as MEMORY_AND_DISK_SER storage level but replicate each partition to two cluster nodes.
// MAGIC
// MAGIC DISK_ONLY – In this storage level, DataFrame is stored only on disk and the CPU computation time is high as I/O is involved.
// MAGIC
// MAGIC DISK_ONLY_2 – Same as DISK_ONLY storage level but replicate each partition to two cluster nodes.
// MAGIC `

// COMMAND ----------

import org.apache.spark.storage.StorageLevel

val rdd1 = words.cache()

val rdd2 = words2.persist(StorageLevel.MEMORY_AND_DISK)

// COMMAND ----------

rdd2.getStorageLevel

// COMMAND ----------

println(words.getCheckpointFile,
words.getNumPartitions,
 words.getResourceProfile,
words.getStorageLevel)

// COMMAND ----------

// MAGIC %md
// MAGIC ###checkpoint
// MAGIC *Checkpointing
// MAGIC is the act of saving an RDD to disk so that future references to this RDD point to those
// MAGIC intermediate partitions on disk rather than recomputing the RDD from its original source. This is
// MAGIC similar to caching except that it’s not stored in memory, only disk.
// MAGIC *

// COMMAND ----------

spark.sparkContext.setCheckpointDir("dbfs:/FileStore/tables/suankiData/checkpointing")
words.checkpoint()

// COMMAND ----------

// MAGIC %fs
// MAGIC
// MAGIC ls dbfs:/FileStore/tables/suankiData/checkpointing

// COMMAND ----------

words.count()

// COMMAND ----------

words.pipe("wc -l").collect()  //Array(9, 9)  so total 18 line ..b/c it results per partitions 

// COMMAND ----------

// MAGIC %md
// MAGIC ####mapPartitions
// MAGIC *Spark operates on a per-partition basis when it comes to
// MAGIC actually executing code. You also might have noticed earlier that the return signature of a map
// MAGIC function on an RDD is actually MapPartitionsRDD. This is because map is just a row-wise alias
// MAGIC for mapPartitions, which makes it possible for you to map an individual partition (represented
// MAGIC as an iterator)*

// COMMAND ----------

//words.map(part => part)  //org.apache.spark.rdd.RDD[String] = MapPartitionsRDD[394]


words.mapPartitions(part => Iterator[Int](1)).sum()

/**
Naturally, this means that we operate on a per-partition basis and allows us to perform an
operation on that entire partition.
**/

// COMMAND ----------

words.mapPartitions(part => part.map(row => (row, row.startsWith("S")))).collect  //doing same thing as  partition by partition

// COMMAND ----------

// in Scala
def indexedFunc(partitionIndex:Int, withinPartIterator: Iterator[String]) = {
      withinPartIterator.toList.map(
      value => s"Partition: $partitionIndex => $value").iterator
}

words.mapPartitionsWithIndex(indexedFunc).collect()



// COMMAND ----------

words.mapPartitionsWithIndex((indx,part) => part.toList.map(row => row).iterator).collect()

// COMMAND ----------

// MAGIC %md
// MAGIC #####foreachPartition
// MAGIC *Although mapPartitions needs a return value to work properly, this next function does not.
// MAGIC foreachPartition simply iterates over all the partitions of the data. The difference is that the
// MAGIC function has no return value. This makes it great for doing something with each partition like
// MAGIC writing it out to a database*

// COMMAND ----------

// MAGIC
// MAGIC %fs
// MAGIC
// MAGIC ls dbfs:/FileStore/tables/suankiData/

// COMMAND ----------

val file = new java.io.File(raw"dbfs:/FileStore/tables/suankiData/temp/random.txt")

val pw = new java.io.PrintWriter(file)
pw.write("test_file")


// COMMAND ----------

val col= "this is test class for big data analytics".split(" ")
val words = spark.sparkContext.parallelize(col,2)

words.foreachPartition { iter =>
  import java.io._
  import scala.util.Random
   
  val randomFileName = new Random().nextInt()

  //  val pw = new PrintWriter(new File(s"dbfs:/FileStore/tables/suankiData/temp/random.txt"))
    val pw = new PrintWriter(new File(s"C:\\Users\\sujee\\OneDrive\\Documents\\bigdata_and_hadoop\\scala\\tmp\\random_$randomFileName.txt"))
   
  while (iter.hasNext) {
      pw.write(iter.next())
    } 

  pw.close()
}

// COMMAND ----------

// MAGIC %md
// MAGIC
// MAGIC ###glom
// MAGIC *glom is an interesting function that takes every partition in your dataset and converts them to
// MAGIC arrays. This can be useful if you’re going to collect the data to the driver and want to have an
// MAGIC array for each partition.*

// COMMAND ----------

words.glom.collect()

// COMMAND ----------

// MAGIC %md
// MAGIC ** org.apache.spark.sql.Row**

// COMMAND ----------

import org.apache.spark.sql.Row


 val row = Row(1, true, "a string", null)  // row: Row = [1,true,a string,null]
 val firstValue = row(0)  // firstValue: Any = 1
 val fourthValue = row(3)  // fourthValue: Any = null

 val firstValueBypremtiveAccess = row.getInt(0)   // firstValue: Int = 1
 val isNull = row.isNullAt(3)    // isNull: Boolean = true

// COMMAND ----------

//can be access using match
spark.sql("SELECT order_id, order_status FROM retail_db.orders").rdd.map {
   case Row(key: Int, value: String) =>  key -> value
 }.toDF().show(2)

// COMMAND ----------

val fullName = Row("Sujeet","Kumar","Singh")

//creating row from seq

val SeqOfData = scala.collection.mutable.Seq("This","is" ,"Test" ,"Sequence", null, 1, 'c')

val RowOfSequence = Row.fromSeq(SeqOfData)

 println(fullName.toSeq) //WrappedArray(1, true, a string, null)

// COMMAND ----------

println(fullName(0)) //sujeet
println(fullName.getString(0)) //
println(RowOfSequence.getAs[String](2))

// COMMAND ----------

spark.sql("SELECT order_id, order_status FROM retail_db.orders").rdd.map{
   case Row(key: Int, value: String) =>  Row(key -> value)
 }.filter(row => ! row.anyNull).count()  //anyNull to test if there are null
